File: lib/labeling-sample.ts
import {
  addToReservoir,
  createSeededRandom,
  shuffleInPlace,
} from "./reservoir-sample.js";

export interface LabelingEmail {
  id: string;
  gmail_message_id: string;
  gmail_thread_id: string;
  internal_date: string;
  direction: "incoming" | "outgoing" | "unknown";
  from_name: string | null;
  from_email: string | null;
  to_recipients: Array<{ name: string | null; email: string | null; raw: string }>;
  subject: string;
  snippet: string;
  body_text: string;
  label_ids: string[];
}

export const LABEL_CATEGORIES = [
  "applied",
  "outreach",
  "reply_needed",
  "interview_assessment",
  "offer",
  "rejected",
  "other",
  "uncertain",
] as const;

const atcPattern = /(^|[^a-z0-9])atc([^a-z0-9]|$)/i;

export function isAtcEmail(email: LabelingEmail): boolean {
  return [
    email.from_name,
    email.from_email,
    ...email.to_recipients.flatMap((recipient) => [
      recipient.name,
      recipient.email,
      recipient.raw,
    ]),
    email.subject,
  ].some((value) => value != null && atcPattern.test(value));
}

export function selectLabelingSample(
  emails: LabelingEmail[],
  sampleSize: number,
  seed: string,
): Array<LabelingEmail & { selection_reason: "atc_required" | "random" }> {
  const required = emails.filter(isAtcEmail);
  if (required.length > sampleSize) {
    throw new Error(
      `${required.length} ATC emails exceed the requested ${sampleSize}-email review size`,
    );
  }

  const random = createSeededRandom(seed);
  const reservoir: LabelingEmail[] = [];
  let randomCandidatesSeen = 0;
  for (const email of emails) {
    if (isAtcEmail(email)) continue;
    randomCandidatesSeen += 1;
    addToReservoir(reservoir, email, randomCandidatesSeen, sampleSize, random);
  }
  shuffleInPlace(reservoir, random);

  const selected = [
    ...required.map((email) => ({ ...email, selection_reason: "atc_required" as const })),
    ...reservoir
      .slice(0, sampleSize - required.length)
      .map((email) => ({ ...email, selection_reason: "random" as const })),
  ];
  shuffleInPlace(selected, random);
  return selected;
}

const discoveryPattern = /\b(reply|respond|response|action required|action needed|next steps?|additional (questions?|information)|right to represent|representation|send (me )?(your )?(updated )?resume|interview|screening|schedule|meeting|call|assessment|test|case study|challenge|exercise|puzzle|offer|decision|rejected|not moving forward)\b/i;

function discoveryCandidate(email: LabelingEmail): boolean {
  if (email.direction === "outgoing") return true;
  return discoveryPattern.test(`${email.subject}\n${email.snippet}`);
}

function reservoirSample<T>(values: T[], count: number, seed: string): T[] {
  const random = createSeededRandom(seed);
  const reservoir: T[] = [];
  values.forEach((value, index) => {
    addToReservoir(reservoir, value, index + 1, count, random);
  });
  shuffleInPlace(reservoir, random);
  return reservoir;
}

export function selectAdditionalSample(
  emails: LabelingEmail[],
  excludedIds: ReadonlySet<string>,
  count: number,
  seed: string,
  strategy: "balanced" | "random",
): LabelingEmail[] {
  const candidates = emails.filter((email) => !excludedIds.has(email.id));
  if (candidates.length < count) {
    throw new Error(
      `Only ${candidates.length} unseen emails remain; cannot add ${count}`,
    );
  }
  if (strategy === "random") return reservoirSample(candidates, count, seed);

  const priority = candidates.filter(discoveryCandidate);
  const priorityTarget = Math.min(priority.length, Math.ceil(count * 0.7));
  const selectedPriority = reservoirSample(priority, priorityTarget, `${seed}:priority`);
  const selectedIds = new Set(selectedPriority.map((email) => email.id));
  const remaining = candidates.filter((email) => !selectedIds.has(email.id));
  const selectedRandom = reservoirSample(
    remaining,
    count - selectedPriority.length,
    `${seed}:remainder`,
  );
  return [...selectedPriority, ...selectedRandom];
}

File: src/email_automation_jev/labeling_sample.py
from __future__ import annotations

import re
from typing import Literal, TypedDict, cast

from .models import AddressValue, Direction
from .reservoir_sample import add_to_reservoir, create_seeded_random, shuffle_in_place


class LabelingEmail(TypedDict):
    id: str
    gmail_message_id: str
    gmail_thread_id: str
    internal_date: str
    direction: Direction
    from_name: str | None
    from_email: str | None
    to_recipients: list[AddressValue]
    subject: str
    snippet: str
    body_text: str
    label_ids: list[str]


class SampledLabelingEmail(LabelingEmail):
    """A labeling email annotated with why it entered the review sample."""

    selection_reason: str


LABEL_CATEGORIES = (
    "applied",
    "outreach",
    "reply_needed",
    "interview_assessment",
    "offer",
    "rejected",
    "other",
    "uncertain",
)
ATC_PATTERN = re.compile(r"(^|[^a-z0-9])atc([^a-z0-9]|$)", re.I)
DISCOVERY_PATTERN = re.compile(
    r"\b(reply|respond|response|action required|action needed|next steps?|"
    r"additional (questions?|information)|right to represent|representation|"
    r"send (me )?(your )?(updated )?resume|interview|screening|schedule|meeting|"
    r"call|assessment|test|case study|challenge|exercise|puzzle|offer|decision|"
    r"rejected|not moving forward)\b",
    re.I,
)


def is_atc_email(email: LabelingEmail) -> bool:
    values: list[str | None] = [email["from_name"], email["from_email"], email["subject"]]
    for recipient in email["to_recipients"]:
        values.extend([recipient.get("name"), recipient.get("email"), recipient.get("raw")])
    return any(value is not None and ATC_PATTERN.search(value) for value in values)


def _reservoir(values: list[LabelingEmail], count: int, seed: str) -> list[LabelingEmail]:
    rng = create_seeded_random(seed)
    result: list[LabelingEmail] = []
    for index, value in enumerate(values, 1):
        add_to_reservoir(result, value, index, count, rng)
    shuffle_in_place(result, rng)
    return result


def select_labeling_sample(emails: list[LabelingEmail], sample_size: int, seed: str) -> list[SampledLabelingEmail]:
    required = [email for email in emails if is_atc_email(email)]
    if len(required) > sample_size:
        raise ValueError(f"{len(required)} ATC emails exceed the requested {sample_size}-email review size")
    rng = create_seeded_random(seed)
    candidates = [email for email in emails if not is_atc_email(email)]
    selected = [cast(SampledLabelingEmail, {**email, "selection_reason": "atc_required"}) for email in required] + [
        cast(SampledLabelingEmail, {**email, "selection_reason": "random"})
        for email in _reservoir(candidates, sample_size - len(required), seed)
    ]
    shuffle_in_place(selected, rng)
    return selected


def select_additional_sample(
    emails: list[LabelingEmail],
    excluded_ids: set[str],
    count: int,
    seed: str,
    strategy: Literal["balanced", "random"],
) -> list[LabelingEmail]:
    candidates = [email for email in emails if email["id"] not in excluded_ids]
    if len(candidates) < count:
        raise ValueError(f"Only {len(candidates)} unseen emails remain; cannot add {count}")
    if strategy == "random":
        return _reservoir(candidates, count, seed)
    priority = [email for email in candidates if email["direction"] == "outgoing" or DISCOVERY_PATTERN.search(f"{email['subject']}\n{email['snippet']}")]
    target = min(len(priority), (count * 7 + 9) // 10)
    selected_priority = _reservoir(priority, target, f"{seed}:priority")
    selected_ids = {email["id"] for email in selected_priority}
    remaining = [email for email in candidates if email["id"] not in selected_ids]
    return selected_priority + _reservoir(remaining, count - len(selected_priority), f"{seed}:remainder")
